src/v2/Carousel/Carousel.sol

```diff
@@ -0,0 +1,521 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import "../VaultV2.sol";
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @author Y2K Finance Team
+
+contract Carousel is VaultV2 {
+    using SafeERC20 for IERC20;
+    using FixedPointMathLib for uint256;
+    /*///////////////////////////////////////////////////////////////
+                               IMMUTABLES AND STORAGE
+    //////////////////////////////////////////////////////////////*/
+    // Earthquake parameters
+    uint256 public relayerFee;
+    uint256 public closingTimeFrame;
+    address public feeTreasury;
+    IERC20 public emissionsToken;
+
+    mapping(address => uint256) public ownerToRollOverQueueIndex;
+    QueueItem[] public rolloverQueue;
+    QueueItem[] public depositQueue;
+    mapping(uint256 => uint256) public rolloverAccounting;
+    mapping(uint256 => mapping(address => uint256)) private _emissionsBalances;
+    mapping(uint256 => uint256) private emissions;
+
+    /*//////////////////////////////////////////////////////////////
+                                 CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    constructor(
+        address _assetAddress,
+        string memory _name,
+        string memory _symbol,
+        string memory _tokenURI,
+        address _token,
+        uint256 _strike,
+        address _controller,
+        address _treasury,
+        address _feeTreasury,
+        address _emissionsToken,
+        uint256 _relayerFee
+    )
+        VaultV2(
+            _assetAddress,
+            _name,
+            _symbol,
+            _tokenURI,
+            _token,
+            _strike,
+            _controller,
+            _treasury
+        )
+    {
+        emissionsToken = IERC20(_emissionsToken);
+        relayerFee = _relayerFee;
+        feeTreasury = _feeTreasury;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        DEPOSIT/WITHDRAWAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function deposit(
+        uint256 _id,
+        uint256 _assets,
+        address _receiver
+    )
+        public
+        override(VaultV2)
+        minRequiredDeposit(_assets)
+        epochIdExists(_id)
+        epochHasNotStarted(_id)
+        nonReentrant
+    {
+        if (_receiver == address(0)) revert AddressZero();
+
+        _asset().safeTransferFrom(msg.sender, address(this), _assets);
+
+        // mint logic, either in queue or direct deposit
+        if (queueClosed(_id)) {
```

Comment:
maybe you want to have lower and higher bound for the queue closed, like when it is `late` and `impossible`